<a href="https://colab.research.google.com/github/fernandotoscano/BIOINFORMATICS/blob/main/CDD_ML_Part_3_Acetylcholinesterase_Descriptor_Dataset_Preparation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Bioinformatics Project - Computational Drug Discovery [Part 3] Descriptor Calculation and Dataset Preparation**

Chanin Nantasenamat

[*'Data Professor' YouTube channel*](http://youtube.com/dataprofessor)

In this Jupyter notebook, we will be building a real-life **data science project** that you can include in your **data science portfolio**. Particularly, we will be building a machine learning model using the ChEMBL bioactivity data.

In **Part 3**, we will be calculating molecular descriptors that are essentially quantitative description of the compounds in the dataset. Finally, we will be preparing this into a dataset for subsequent model building in Part 4.

---

**THIS IS THE CONTINUATION OF "Cópia_de_CDD_ML_Part_2_Exploratory_Data_Analysis_colab_(2).ipynb"  [52:16]**,
**IN THIS PART_3 WE'LL BE CALCULATING MOLECULAR DESCRIPTORS AND THEN WE'RE GOING TO PREPARE THE DATA SET WHICH WILL BE USED FOR THE NEXT PART, PART_4.
AND IN PART_4 WE'RE GONNA PERFORM SOME MODEL BUILDING, SO WE ARE GOING TO NEED TO DOWNLOAD THE PaDEL-Descriptor, SOFTWARE WHICH IS PROVIDED IN THE GITHUB OF DATAPROFESSOR, AND WE'LL HAVE ACCESS TO THE LINK OF THE ORIGINAL WEBSITE OF THE DEVELOPERS OF THE "PAaDEL" DESCRIPTORS ALONG WITH THE LINK OF THE ORIGINAL RESEARCH PAPER **

## **Download PaDEL-Descriptor**

In [1]:
! wget https://github.com/dataprofessor/bioinformatics/raw/master/padel.zip
! wget https://github.com/dataprofessor/bioinformatics/raw/master/padel.sh

--2025-01-24 11:19:58--  https://github.com/dataprofessor/bioinformatics/raw/master/padel.zip
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/dataprofessor/bioinformatics/master/padel.zip [following]
--2025-01-24 11:19:59--  https://raw.githubusercontent.com/dataprofessor/bioinformatics/master/padel.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25768637 (25M) [application/zip]
Saving to: ‘padel.zip’

padel.zip           100%[===================>]  24.57M   133MB/s    in 0.2s    

2025-01-24 11:19:59 (133 MB/s) - ‘padel.zip’ saved [25768637/25768637]

--2025-01-24 11:19:59--  https://github

**SO THE "/padel.zip" has been downloded along the "/padel.sh" file which is the shell script file containing instructions on how to run the padle calculation:
"! wget https://github.com/dataprofessor/bioinformatics/raw/master/padel.zip
! wget https://github.com/dataprofessor/bioinformatics/raw/master/padel.sh"
because here we're going to use "padle" to calculate the molecular descriptors  **

**SO HERE, WE'RE GOING TO UNZIP THE THE FOLDER:**

In [2]:
! unzip padel.zip

Archive:  padel.zip
   creating: PaDEL-Descriptor/
  inflating: __MACOSX/._PaDEL-Descriptor  
  inflating: PaDEL-Descriptor/MACCSFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._MACCSFingerprinter.xml  
  inflating: PaDEL-Descriptor/AtomPairs2DFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._AtomPairs2DFingerprinter.xml  
  inflating: PaDEL-Descriptor/EStateFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._EStateFingerprinter.xml  
  inflating: PaDEL-Descriptor/Fingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._Fingerprinter.xml  
  inflating: PaDEL-Descriptor/.DS_Store  
  inflating: __MACOSX/PaDEL-Descriptor/._.DS_Store  
   creating: PaDEL-Descriptor/license/
  inflating: __MACOSX/PaDEL-Descriptor/._license  
  inflating: PaDEL-Descriptor/KlekotaRothFingerprintCount.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._KlekotaRothFingerprintCount.xml  
  inflating: PaDEL-Descriptor/config  
  inflating: __MACOSX/PaDEL-Descriptor/._config  
  inf

## **Load bioactivity data**

Download the curated ChEMBL bioactivity data that has been pre-processed from Parts 1 and 2 of this Bioinformatics Project series. Here we will be using the **bioactivity_data_3class_pIC50.csv** file that essentially contain the pIC50 values that we will be using for building a regression model.

**SO WE'RE GOING TO DOWNLOAD THE FILE "acetylcholinesterase_04_bioactivity_data_3class_pIC50.csv", WHICH CONTAIN THE "pIC50" ALONG WITH THE 3 BIOACTIVITY CLASS. SO WE'RE GOING TO IMPORT PANDAS AS PD AFTER THIS:**

In [3]:
! wget https://raw.githubusercontent.com/dataprofessor/data/master/acetylcholinesterase_04_bioactivity_data_3class_pIC50.csv

--2025-01-24 11:20:00--  https://raw.githubusercontent.com/dataprofessor/data/master/acetylcholinesterase_04_bioactivity_data_3class_pIC50.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 655414 (640K) [text/plain]
Saving to: ‘acetylcholinesterase_04_bioactivity_data_3class_pIC50.csv’

acetylcholinesteras 100%[===================>] 640.05K  --.-KB/s    in 0.05s   

2025-01-24 11:20:00 (11.6 MB/s) - ‘acetylcholinesterase_04_bioactivity_data_3class_pIC50.csv’ saved [655414/655414]



In [4]:
import pandas as pd

**AFTER IMPORT THE PANDAS AS PD, WE'VE TO IMPORT THE .csv FILE, AND ASSIGN IT THE "df3" VARIABLE NAME:**

In [5]:
df3 = pd.read_csv('acetylcholinesterase_04_bioactivity_data_3class_pIC50.csv')

In [6]:
df3

Unnamed: 0 molecule_chembl_id  \
0              0       CHEMBL133897   
1              1       CHEMBL336398   
2              2       CHEMBL131588   
3              3       CHEMBL130628   
4              4       CHEMBL130478   
...          ...                ...   
4690        4690      CHEMBL4293155   
4691        4691      CHEMBL4282558   
4692        4692      CHEMBL4281727   
4693        4693      CHEMBL4292349   
4694        4694      CHEMBL4278260   

                                       canonical_smiles         class  \
0                 CCOc1nn(-c2cccc(OCc3ccccc3)c2)c(=O)o1        active   
1            O=C(N1CCCCC1)n1nc(-c2ccc(Cl)cc2)nc1SCC1CC1        active   
2     CN(C(=O)n1nc(-c2ccc(Cl)cc2)nc1SCC(F)(F)F)c1ccccc1      inactive   
3         O=C(N1CCCCC1)n1nc(-c2ccc(Cl)cc2)nc1SCC(F)(F)F        active   
4             CSc1nc(-c2ccc(OC(F)(F)F)cc2)nn1C(=O)N(C)C        active   
...                                                 ...           ...   
4690  CC(C)(C)c1cc(/C=C/C(=O)NCCC2CCN(Cc3ccccc3Cl)CC...  intermediate   
4691  CC(C)(C)c1cc(/C=C/C(=O)NCCC2CCN(Cc3cccc(Cl)c3)...  intermediate   
4692  CC(C)(C)c1cc(/C=C/C(=O)NCCC2CCN(Cc3ccc(Br)cc3)...  intermediate   
4693  CC(C)(C)c1cc(/C=C/C(=O)NCCC2CCN(Cc3cccc([N+](=...  intermediate   
4694  CC(C)(C)c1cc(/C=C/C(=O)NCCC2CCN(Cc3ccc(C#N)cc3...  intermediate   

           MW     LogP  NumHDonors  NumHAcceptors     pIC50  
0     312.325  2.80320         0.0            6.0  6.124939  
1     376.913  4.55460         0.0            5.0  7.000000  
2     426.851  5.35740         0.0            5.0  4.301030  
3     404.845  4.70690         0.0            5.0  6.522879  
4     346.334  3.09530         0.0            6.0  6.096910  
...       ...      ...         ...            ...       ...  
4690  511.150  7.07230         2.0            3.0  5.612610  
4691  511.150  7.07230         2.0            3.0  5.595166  
4692  555.601  7.18140         2.0            3.0  5.419075  
4693  521.702  6.32710         2.0            5.0  5.460924  
4694  501.715  6.29058         2.0            4.0  5.555955  

[4695 rows x 9 columns]

**IN  THE NEXT LINE OF CODE WE'RE GOING TO SELECRT ONLY THE "canonical_smiles" COLUMN ALONG WITH THE "molecule_chembl_id" COLUMN, AND WE'RE GOING TO PUT IT IN THE SELECTION VARIABLE AND THEN WE'RE GOING TO SUBSET THE DATA BY USING "df3[selection]", WHICH CONTAINS THE NAME OF THE PRECISE COLUMNS THAT WE WANTED AND THEN WE'RE GOING TO ASSIGN THE NAME OF "df3_selection" AND THEN WE'RE GONNA SAVE IT AS "molecule.smi" (THE COMPLETE LINE OF CODE: "df3_selection.to_csv('molecule.smi', sep='\t', index=False, header=False)" ), SO LET'S RUN IT: **

In [7]:
selection = ['canonical_smiles','molecule_chembl_id']
df3_selection = df3[selection]
df3_selection.to_csv('molecule.smi', sep='\t', index=False, header=False)

** *LET'S LOOK AT THE FILE USING THE "bash" (explanation: " !cat " magic in order to view the structure (e.g. number of columns) of a huge file (150 MB). When I simply execute !cat filename.csv, Jupyter hangs. Is it possible to specify how many rows to display?". "bash" The 'Cat Command' is a command in Computer Science that is used to output or view the contents of a file. It can also be utilized to join or concatenate multiple files together.).
SO IT CONTAINS THE SMILES NOTATION HERE AND THE NAME OF THE MOLECULE.
SO A QUICK RECAP, THESE SMILES NOTATION REPRESENTS THE CHEMICAL INFORMATION THAT RETAIN THE CHEMICAL STRUCTURE, SO THE "C" REPREASENTS THE CARBON ATOM, THE "O" REPRESENTS THE OXIGEN, AND "N" THE NYTROGEN. SO LET'S CONTINUE: **

In [8]:
! cat molecule.smi | head -5

CCOc1nn(-c2cccc(OCc3ccccc3)c2)c(=O)o1	CHEMBL133897
O=C(N1CCCCC1)n1nc(-c2ccc(Cl)cc2)nc1SCC1CC1	CHEMBL336398
CN(C(=O)n1nc(-c2ccc(Cl)cc2)nc1SCC(F)(F)F)c1ccccc1	CHEMBL131588
O=C(N1CCCCC1)n1nc(-c2ccc(Cl)cc2)nc1SCC(F)(F)F	CHEMBL130628
CSc1nc(-c2ccc(OC(F)(F)F)cc2)nn1C(=O)N(C)C	CHEMBL130478


**IN THE NEXT LINE OF CODE, WE'RE GOING TO SEE HOW MANY LINES FO MOLECULES DO WE HAVE, AND WE HAVE 4695 LINES, WHICH MATCHES WITH THE NUMBER OF ROWS OF "df3" above. SO WE WANTED TO CHECK THAT ALL OF THE ROWS ARE COMING IN THE "molecule.smi" FILE, OK?**

In [9]:
! cat molecule.smi | wc -l

4695


## **Calculate fingerprint descriptors**


### **Calculate PaDEL descriptors**

**SO WE'RE GONNA PERFORM THE DESCRIPTOR CALCULATION BY RUNNING BASH "! cat padel.sh".
SO MAYBE WE'RE WONDERING WHAT IS INSIDE THE "padel.sh" FILE, LET'S HAVE A LOOK:**

In [10]:
! cat padel.sh

java -Xms1G -Xmx1G -Djava.awt.headless=true -jar ./PaDEL-Descriptor/PaDEL-Descriptor.jar -removesalt -standardizenitro -fingerprints -descriptortypes ./PaDEL-Descriptor/PubchemFingerprinter.xml -dir ./ -file descriptors_output.csv


**SO THE FILE CONTAINS THE INNSTRUTCTIONS SO WE'RE GOING TO USE JAVA AND THEN WE'RE GOING USE 1 GIGABYTE OF MEMORY. [55:14] AND WE'RE GONNA USE THIS OPTION  "-Djava.awt.headless=true" BECAUSE WE DON'T HAVE A DISPLAY ON THE GOOGLE COLAB SO WE'RE GOING TO USE THE JAVA "-Djava.awt.headless=true" AND THEN WE'RE GOING SPECIFY THE JAR FUNCTION BECAUSE WE'RE GOING TO USE THE "PaDEL-Descriptor.jar" FILE AND THEN HER WE'RE GONNA REMOVE THE "SALT" "-removesalt" AND SALTS ARE SODIUM AND CHLORIDE WHICH ARE IN CHEMICAL STRUCTURE AND SOTHIS PROGRAM WILL AUTOMATICALLY REMOVE ALL SALTS AND ALSO SMALL ORGANIC ACID FROM THE CHEMICAL STRUCTURE SO IF THAT SOUND GIBBERISH (SOA A DISPARATE), THEN SO IT ESSENTIALLY MEANS THAT WE ARE CLEANING THE CHEMICAL STRUCTURE SO THAT THERE ARE NO IMPURITY. AND HERE ARE THE OTHER OPTIONS RELATED TO HOW WE ALSO CLEAN THE CHEMICAL STRUCTURE AND THEN THIS OPTION "-fingerprints" TELLS THE PROGRAM THAT WE'RE GOING TO COMPUTE THE MOLECULAR FINGERPRINT AND THE FINGERPRINT TYPE "-fingerprints -descriptortypes" WILL BE PUBCHEM "PubchemFingerprinter", AND FINALLY IT WILL OUTPUT THE DESCRIPTORS INTO THE FILE "-file descriptors_output.csv". AND LET'S RUN IT:**

In [11]:
# ! bash padel.sh

**SOME MOLECULES TAKING SOME SECONDS AN AVERAGE OF 0.28 SECONDS. BECAUSE WE HAVE 4695 MOLECULES IT TOKE ABOUT 21 MINUTES. WE COULD IF WE WANT DOWNLOAD DIRECTLY THE COMPUTED FILE WHICH IS IN DATAPROFESSOR/DATA WITH THE NAME "descriptors_output.csv", SO WHY DON'T WE HEAD OVER TO AND DOWNLOAD IT TO OUR COMPUTER. AND THEN WE'LL GONNA CHANGE THE TYPE (IF IT ISN'T IN THE FORMAT .csv, WE HAVE SAVE IT AS .csv, BECAUSE SOME FILES ARE IN THE FORMAT .txt, WHEN WE HAVE THE OPTION SAVE AS .csv).  
SO IN CASE IF WE DON'T WANT TO RUN THE CODE ABOVE WE CAN UPLOAD THE FILE "descriptors_output.csv", AND NEXT WE'LL LIST THE FILES THAT WE HAVE UPLOADED (I'VE DEACTIVATED THE CODE LINE ABOVE):     

In [12]:
! ls -l

total 34296
-rw-r--r-- 1 root root   655414 Jan 24 11:20 acetylcholinesterase_04_bioactivity_data_3class_pIC50.csv
-rw-r--r-- 1 root root  8356701 Jan 24 11:18 descriptors_output.csv
drwxr-xr-x 3 root root     4096 Jan 24 11:20 __MACOSX
-rw-r--r-- 1 root root   310658 Jan 24 11:20 molecule.smi
drwxrwxr-x 4 root root     4096 May 30  2020 PaDEL-Descriptor
-rw-r--r-- 1 root root      231 Jan 24 11:20 padel.sh
-rw-r--r-- 1 root root 25768637 Jan 24 11:19 padel.zip
drwxr-xr-x 1 root root     4096 Jan 22 14:23 sample_data


**SO NOTE THAT THE FILE SIZE "8356701" SO IT'S ABOUT 8.3 MEGABYTES.**  

## **Preparing the X and Y Data Matrices**

### **X data matrix**

**AND SO WE'RE GONNA READ THE DESCRIPTORS OUTPUT INTO "df3_X", AND SO HERE WE'RE GOING TO PREPARE THE X, AND "Y" DATA MATRICES. AND THE "X" DATA WILL COMPRICE OF THE MOLECULAR DESCRIPTORS WHICH ARE THE PUBCHEM FINGRERPRINTS SO LET'S HAVE A LOOK:**

In [13]:
df3_X = pd.read_csv('descriptors_output.csv')

In [14]:
df3_X

Name  PubchemFP0  PubchemFP1  PubchemFP2  PubchemFP3  \
0      CHEMBL336398           1           1           1           0   
1      CHEMBL133897           1           1           1           0   
2      CHEMBL130628           1           1           1           0   
3      CHEMBL131588           1           1           0           0   
4      CHEMBL130478           1           1           0           0   
...             ...         ...         ...         ...         ...   
4690  CHEMBL4293155           1           1           1           1   
4691  CHEMBL4282558           1           1           1           1   
4692  CHEMBL4281727           1           1           1           1   
4693  CHEMBL4292349           1           1           1           1   
4694  CHEMBL4278260           1           1           1           1   

      PubchemFP4  PubchemFP5  PubchemFP6  PubchemFP7  PubchemFP8  ...  \
0              0           0           0           0           0  ...   
1              0           0           0           0           0  ...   
2              0           0           0           0           0  ...   
3              0           0           0           0           0  ...   
4              0           0           0           0           0  ...   
...          ...         ...         ...         ...         ...  ...   
4690           0           0           0           0           0  ...   
4691           0           0           0           0           0  ...   
4692           0           0           0           0           0  ...   
4693           0           0           0           0           0  ...   
4694           0           0           0           0           0  ...   

      PubchemFP871  PubchemFP872  PubchemFP873  PubchemFP874  PubchemFP875  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
4690             0             0             0             0             0   
4691             0             0             0             0             0   
4692             0             0             0             0             0   
4693             0             0             0             0             0   
4694             0             0             0             0             0   

      PubchemFP876  PubchemFP877  PubchemFP878  PubchemFP879  PubchemFP880  
0                0             0             0             0             0  
1                0             0             0             0             0  
2                0             0             0             0             0  
3                0             0             0             0             0  
4                0             0             0             0             0  
...            ...           ...           ...           ...           ...  
4690             0             0             0             0             0  
4691             0             0             0             0             0  
4692             0             0             0             0             0  
4693             0             0             0             0             0  
4694             0             0             0             0             0  

[4695 rows x 882 columns]

**AND SO WE'RE GOING TO DELETE THE 1ST COLUMN HERE THE "NAME" BECAUSE WE WANT ONLY THE MOLECULAR FEATURES SO LET'S DROP IT USING THE DOT DROP FUNCTION (df3_X = df3_X.drop(colummns=['Name'])", AND THE NAME OF THE COLUMN THAT WE WANT IT TO BE DROPED AND THEN WE CAN SEE THAT THE "Name" COLUMN HAS BEEN DROPPED (AND THEN WE REASSIGN IT BACK TO "df3_X" ): **

In [15]:
df3_X = df3_X.drop(columns=['Name'])
df3_X

PubchemFP0  PubchemFP1  PubchemFP2  PubchemFP3  PubchemFP4  PubchemFP5  \
0              1           1           1           0           0           0   
1              1           1           1           0           0           0   
2              1           1           1           0           0           0   
3              1           1           0           0           0           0   
4              1           1           0           0           0           0   
...          ...         ...         ...         ...         ...         ...   
4690           1           1           1           1           0           0   
4691           1           1           1           1           0           0   
4692           1           1           1           1           0           0   
4693           1           1           1           1           0           0   
4694           1           1           1           1           0           0   

      PubchemFP6  PubchemFP7  PubchemFP8  PubchemFP9  ...  PubchemFP871  \
0              0           0           0           1  ...             0   
1              0           0           0           1  ...             0   
2              0           0           0           1  ...             0   
3              0           0           0           1  ...             0   
4              0           0           0           1  ...             0   
...          ...         ...         ...         ...  ...           ...   
4690           0           0           0           1  ...             0   
4691           0           0           0           1  ...             0   
4692           0           0           0           1  ...             0   
4693           0           0           0           1  ...             0   
4694           0           0           0           1  ...             0   

      PubchemFP872  PubchemFP873  PubchemFP874  PubchemFP875  PubchemFP876  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
4690             0             0             0             0             0   
4691             0             0             0             0             0   
4692             0             0             0             0             0   
4693             0             0             0             0             0   
4694             0             0             0             0             0   

      PubchemFP877  PubchemFP878  PubchemFP879  PubchemFP880  
0                0             0             0             0  
1                0             0             0             0  
2                0             0             0             0  
3                0             0             0             0  
4                0             0             0             0  
...            ...           ...           ...           ...  
4690             0             0             0             0  
4691             0             0             0             0  
4692             0             0             0             0  
4693             0             0             0             0  
4694             0             0             0             0  

[4695 rows x 881 columns]

## **Y variable**

[59:14] **SO NOW LET'S CREATE THE "Y" DATA MATRIX (THE DEPENDENT VARIABLE), AND HERE WE'RE GOING TO TAKE THE "pIC50" COLUMN DIRECTLY FROM THE "df3" DATAFRAME WHICH IS THE INITIAL LOADED DATA FRAME AND THEN WE'RE GOING TO ASSIGN IT TO THE "df3_Y":**

### **Convert IC50 to pIC50**

In [16]:
df3_Y = df3['pIC50']
df3_Y

0       6.124939
1       7.000000
2       4.301030
3       6.522879
4       6.096910
          ...   
4690    5.612610
4691    5.595166
4692    5.419075
4693    5.460924
4694    5.555955
Name: pIC50, Length: 4695, dtype: float64

## **Combining X and Y variable**

**AND THEN HERE WE'RE GOING TO COMBINE "X" AND "Y" TOGETHER SO THIS REALLY DEPENDS BUT FOR PORTABILITY WE'RE GONNA COMBINEIT AND WE'RE GONNA OUTUP IT INTO A ".csv" FILE WHICH WE WILL BE UPLOADING TO GITHUB, AND THEN WE'RE GONNA USE THAT FOR PART_4:**

In [17]:
dataset3 = pd.concat([df3_X,df3_Y], axis=1)
dataset3

PubchemFP0  PubchemFP1  PubchemFP2  PubchemFP3  PubchemFP4  PubchemFP5  \
0              1           1           1           0           0           0   
1              1           1           1           0           0           0   
2              1           1           1           0           0           0   
3              1           1           0           0           0           0   
4              1           1           0           0           0           0   
...          ...         ...         ...         ...         ...         ...   
4690           1           1           1           1           0           0   
4691           1           1           1           1           0           0   
4692           1           1           1           1           0           0   
4693           1           1           1           1           0           0   
4694           1           1           1           1           0           0   

      PubchemFP6  PubchemFP7  PubchemFP8  PubchemFP9  ...  PubchemFP872  \
0              0           0           0           1  ...             0   
1              0           0           0           1  ...             0   
2              0           0           0           1  ...             0   
3              0           0           0           1  ...             0   
4              0           0           0           1  ...             0   
...          ...         ...         ...         ...  ...           ...   
4690           0           0           0           1  ...             0   
4691           0           0           0           1  ...             0   
4692           0           0           0           1  ...             0   
4693           0           0           0           1  ...             0   
4694           0           0           0           1  ...             0   

      PubchemFP873  PubchemFP874  PubchemFP875  PubchemFP876  PubchemFP877  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
4690             0             0             0             0             0   
4691             0             0             0             0             0   
4692             0             0             0             0             0   
4693             0             0             0             0             0   
4694             0             0             0             0             0   

      PubchemFP878  PubchemFP879  PubchemFP880     pIC50  
0                0             0             0  6.124939  
1                0             0             0  7.000000  
2                0             0             0  4.301030  
3                0             0             0  6.522879  
4                0             0             0  6.096910  
...            ...           ...           ...       ...  
4690             0             0             0  5.612610  
4691             0             0             0  5.595166  
4692             0             0             0  5.419075  
4693             0             0             0  5.460924  
4694             0             0             0  5.555955  

[4695 rows x 882 columns]

In [18]:
dataset3.to_csv('acetylcholinesterase_06_bioactivity_data_3class_pIC50_pubchem_fp.csv', index=False)

# **Let's download the CSV file to your local computer for the Part 3B (Model Building).**

**SO WE'RE GONNA OUTPUT THIS "dataset_3" DATAFRAME INTO A ".csv" FILE, AND NOTICE THAT THE NAME HERE IS FAIRLY LONG AND THE PURPOSRE FOR HAVING SUCH A LONG NAME IS TO ALLOW LESS CONFUSION AND TO ALLOW US TO EASILY SEE WHAT IS THE PURPOSE OF THIS FILE AND SO, THE 1ST SEGMENT OF THIS NAME IS RTHE NAME OF THE TARGET PROTEIN WHICH IS THE "acetylcholinesterase_06", "06" IS JUST THE SEQUENTIAL NUMBER AND SO "_bioactivity_data_3class_pIC50_" ESSSENTIALLY TELL US THAT IT CONTAINS THE BIOACTIVITY DATA INFORMATION ALONG WITH 3 CATEGORICAL CLASS COMPRISING OF ACTIVE  AND INACTIVE AND INTERMEDIATE AND IT ALSO CONTAINS THE "pIC50" VALUES AND THEN [1:00:22] THE LAST SEGMENT HERE IS "_pubchem_fp", SIGNIFYING THAT IT CONTAINS THE "PUBCHEM" FINGERPRINT SO THIS WILL BECOME HANDY WHEN WE HAVE MORE THAN ONE FINGERPRINT TYPE AND SO "PADEL" ALLOWS US TO COMPUTE MORE THEN 10 DIFFERENT FINGERPRINT TYPES. AND SO LET'S MAKE IT OUR HOMEWORK TO TRY TO COMPUTE OTHER DIFFERENT FINGERPRINT TYPES, SO WE WANT TO PLAY  AROUND WITH THE OPTIONS AND SEE WHAT OTHER MOLECULAR FINGERPRINTS ARE AVAILABLE AND THEN WE COULD RENAME THE FILE ACCORDINGLY (EM CONSONÂNCIA/EM CONFORMIDADE). SO LET'S SEE IF WE HAVE ALREADY WRITTEN OUT THE FILE AND IT'S RIHT THERE "acetylcholinesterase_06_bioactivity_data_3class_pIC50_pubchem_fp.csv" ON THE RIGHT PANEL () AND WE HAVE ALREADY DOWNLOADED TO OUR COMPUTER. IF WE OPPENED THE FILE WE CAN SEE THE FINGERPRINTS OF THE PUBCHEM AND IN THE LAST COLUMN IS THE "pIC50". SO WE'RE GOING TO USE THIS FILE TO PERFORM MODEL BUILDING IN THE NEXT CHAPTER.**

# **END OF PART_3**


# **PART_4**
** NOW WELCOME TO PART_4, OF BIOINFORMATICS FROM SCRATCH SERIES WHERE WE 'LL SHOW HOW TO DO A BIOINFORMATICS PROJECT USING MACHINE LEARNING IN A STEP BY STEP**